File: apps/api/internal/hrm/domain/provisioning.go
package domain

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
)

// Provisioning errors
var (
	ErrDuplicatePendingRequest    = errors.New("DUPLICATE_PENDING_REQUEST")
	ErrRequestNotFound            = errors.New("REQUEST_NOT_FOUND")
	ErrInvalidRoleForProvisioning = errors.New("INVALID_ROLE_FOR_PROVISIONING")
	ErrRequestExpired             = errors.New("REQUEST_EXPIRED")
	ErrRequestAlreadyExecuted     = errors.New("REQUEST_ALREADY_EXECUTED")
	ErrOffboardingNotFound        = errors.New("OFFBOARDING_NOT_FOUND")
	ErrInvalidRequestStatus       = errors.New("INVALID_REQUEST_STATUS")
)

// ProvisioningRequest maps to user_provisioning_requests.
type ProvisioningRequest struct {
	ID                    uuid.UUID
	EmployeeID            uuid.UUID
	RequestedBy           uuid.UUID
	RequestedRole         string
	RequestedBranchID     *uuid.UUID
	Status                string
	ApprovalLevel         int16
	BranchApproverID      *uuid.UUID
	BranchApprovedAt      *time.Time
	BranchRejectionReason *string
	HRApproverID          *uuid.UUID
	HRApprovedAt          *time.Time
	HRRejectionReason     *string
	ExecutedBy            *uuid.UUID
	ExecutedAt            *time.Time
	IsEmergency           bool
	EmergencyReason       *string
	Notes                 *string
	ExpiresAt             time.Time
	CreatedAt             time.Time
	UpdatedAt             time.Time
}

// OffboardingChecklist maps to offboarding_checklists.
type OffboardingChecklist struct {
	ID            uuid.UUID
	EmployeeID    uuid.UUID
	ChecklistType string
	InitiatedBy   uuid.UUID
	TargetDate    *time.Time
	Items         []byte
	Status        string
	CompletedAt   *time.Time
	Notes         *string
	CreatedAt     time.Time
	UpdatedAt     time.Time
}

// ─── Params ───────────────────────────────────────────────────────────────────

type CreateProvisioningParams struct {
	EmployeeID        uuid.UUID
	RequestedBy       uuid.UUID
	RequestedRole     string
	RequestedBranchID *uuid.UUID
	IsEmergency       bool
	EmergencyReason   *string
	Notes             *string
}

type ListProvisioningFilter struct {
	Status     string
	EmployeeID *uuid.UUID
	Page       int
	Size       int
}

type BranchApproveParams struct {
	RequestID  uuid.UUID
	ApproverID uuid.UUID
}

type HRApproveParams struct {
	RequestID  uuid.UUID
	ApproverID uuid.UUID
}

type RejectParams struct {
	RequestID  uuid.UUID
	RejectedBy uuid.UUID
	Reason     string
}

type ExecuteParams struct {
	RequestID  uuid.UUID
	ExecutorID uuid.UUID
}

// ExecuteResult is returned after atomic user provisioning.
type ExecuteResult struct {
	Request      *ProvisioningRequest
	UserID       uuid.UUID
	TempPassword string // plain text, returned once; never persisted
}

type CreateOffboardingParams struct {
	EmployeeID    uuid.UUID
	ChecklistType string
	InitiatedBy   uuid.UUID
	TargetDate    *time.Time
	Notes         *string
}

type UpdateOffboardingItemParams struct {
	ChecklistID uuid.UUID
	ItemKey     string
	Completed   bool
	CompletedBy *uuid.UUID
	Notes       *string
}

type ListOffboardingFilter struct {
	Status     string
	EmployeeID *uuid.UUID
	Page       int
	Size       int
}

// ─── Repository interfaces ────────────────────────────────────────────────────

// ProvisioningExpiredAlert is returned by ListExpiredPending for the HRM reminder job.
type ProvisioningExpiredAlert struct {
	RequestID   uuid.UUID
	EmployeeID  uuid.UUID
	RequestedBy uuid.UUID // notify the submitter
	ExpiresAt   time.Time
}

// ProvisioningRepository defines data access for user_provisioning_requests.
type ProvisioningRepository interface {
	Create(ctx context.Context, p CreateProvisioningParams) (*ProvisioningRequest, error)
	FindByID(ctx context.Context, id uuid.UUID) (*ProvisioningRequest, error)
	List(ctx context.Context, f ListProvisioningFilter) ([]*ProvisioningRequest, int64, error)
	HasPendingForEmployee(ctx context.Context, employeeID uuid.UUID) (bool, error)
	BranchApprove(ctx context.Context, p BranchApproveParams) (*ProvisioningRequest, error)
	BranchReject(ctx context.Context, p RejectParams) (*ProvisioningRequest, error)
	HRApprove(ctx context.Context, p HRApproveParams) (*ProvisioningRequest, error)
	HRReject(ctx context.Context, p RejectParams) (*ProvisioningRequest, error)
	Cancel(ctx context.Context, requestID, callerID uuid.UUID) (*ProvisioningRequest, error)
	MarkExecuted(ctx context.Context, requestID, executorID uuid.UUID) (*ProvisioningRequest, error)
	// ListExpiredPending returns PENDING requests whose expires_at has passed.
	// Used by the HRM daily reminder job to notify the requester.
	ListExpiredPending(ctx context.Context) ([]ProvisioningExpiredAlert, error)
}

// OffboardingRepository defines data access for offboarding_checklists.
type OffboardingRepository interface {
	Create(ctx context.Context, p CreateOffboardingParams) (*OffboardingChecklist, error)
	FindByID(ctx context.Context, id uuid.UUID) (*OffboardingChecklist, error)
	List(ctx context.Context, f ListOffboardingFilter) ([]*OffboardingChecklist, int64, error)
	UpdateItem(ctx context.Context, p UpdateOffboardingItemParams) (*OffboardingChecklist, error)
	Complete(ctx context.Context, checklistID, callerID uuid.UUID) (*OffboardingChecklist, error)
}

File: apps/web/src/services/hrm/provisioning.ts
import { api } from '@/lib/api';
import type { PaginatedResult } from '@/types';

// ─── Types ────────────────────────────────────────────────────────────────────

export interface ProvisioningRequest {
  id: string;
  employee_id: string;
  requested_by: string;
  requested_role: string;
  requested_branch_id?: string;
  status: 'PENDING' | 'APPROVED' | 'REJECTED' | 'EXECUTED' | 'CANCELLED';
  approval_level: number;
  branch_approver_id?: string;
  branch_approved_at?: string;
  branch_rejection_reason?: string;
  hr_approver_id?: string;
  hr_approved_at?: string;
  hr_rejection_reason?: string;
  executed_by?: string;
  executed_at?: string;
  is_emergency: boolean;
  emergency_reason?: string;
  notes?: string;
  expires_at: string;
  created_at: string;
  updated_at: string;
}

export interface ExecuteProvisioningResponse {
  request: ProvisioningRequest;
  user_id: string;
  temp_password: string;
}

export interface OffboardingChecklist {
  id: string;
  employee_id: string;
  checklist_type: 'ONBOARDING' | 'OFFBOARDING';
  initiated_by: string;
  target_date?: string;
  items: OffboardingItems;
  status: 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED';
  completed_at?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface OffboardingItem {
  key: string;
  label: string;
  completed: boolean;
  notes?: string;
}

export interface OffboardingItems {
  items: OffboardingItem[];
}

// ─── Request types ─────────────────────────────────────────────────────────────

export interface CreateProvisioningRequest {
  employee_id: string;
  requested_role: string;
  requested_branch_id?: string;
  is_emergency?: boolean;
  emergency_reason?: string;
  notes?: string;
}

export interface RejectRequest {
  reason: string;
}

export interface ExecuteRequest {
  full_name: string;
  email: string;
}

export interface CreateOffboardingRequest {
  employee_id: string;
  checklist_type: 'ONBOARDING' | 'OFFBOARDING';
  target_date?: string;
  notes?: string;
}

export interface UpdateOffboardingItemRequest {
  completed: boolean;
  notes?: string;
}

export interface ListProvisioningParams {
  page?: number;
  size?: number;
  status?: string;
  employee_id?: string;
}

// ─── Services ─────────────────────────────────────────────────────────────────

export const provisioningService = {
  list: (params?: ListProvisioningParams) =>
    api.get<PaginatedResult<ProvisioningRequest>>('/hrm/user-provisioning-requests', { params })
       .then(r => r.data),

  getById: (id: string) =>
    api.get<{ data: ProvisioningRequest }>(`/hrm/user-provisioning-requests/${id}`)
       .then(r => r.data.data),

  create: (data: CreateProvisioningRequest) =>
    api.post<{ data: ProvisioningRequest }>('/hrm/user-provisioning-requests', data)
       .then(r => r.data.data),

  branchApprove: (id: string) =>
    api.post<{ data: ProvisioningRequest }>(`/hrm/user-provisioning-requests/${id}/branch-approve`)
       .then(r => r.data.data),

  branchReject: (id: string, data: RejectRequest) =>
    api.post<{ data: ProvisioningRequest }>(`/hrm/user-provisioning-requests/${id}/branch-reject`, data)
       .then(r => r.data.data),

  hrApprove: (id: string) =>
    api.post<{ data: ProvisioningRequest }>(`/hrm/user-provisioning-requests/${id}/hr-approve`)
       .then(r => r.data.data),

  hrReject: (id: string, data: RejectRequest) =>
    api.post<{ data: ProvisioningRequest }>(`/hrm/user-provisioning-requests/${id}/hr-reject`, data)
       .then(r => r.data.data),

  execute: (id: string, data: ExecuteRequest) =>
    api.post<{ data: ExecuteProvisioningResponse }>(`/hrm/user-provisioning-requests/${id}/execute`, data)
       .then(r => r.data.data),

  cancel: (id: string) =>
    api.post<{ data: ProvisioningRequest }>(`/hrm/user-provisioning-requests/${id}/cancel`)
       .then(r => r.data.data),
};

export const offboardingService = {
  list: (params?: { page?: number; size?: number; status?: string; employee_id?: string }) =>
    api.get<PaginatedResult<OffboardingChecklist>>('/hrm/offboarding', { params })
       .then(r => r.data),

  getById: (id: string) =>
    api.get<{ data: OffboardingChecklist }>(`/hrm/offboarding/${id}`)
       .then(r => r.data.data),

  create: (data: CreateOffboardingRequest) =>
    api.post<{ data: OffboardingChecklist }>('/hrm/offboarding', data)
       .then(r => r.data.data),

  updateItem: (id: string, key: string, data: UpdateOffboardingItemRequest) =>
    api.put<{ data: OffboardingChecklist }>(`/hrm/offboarding/${id}/items/${key}`, data)
       .then(r => r.data.data),

  complete: (id: string) =>
    api.post<{ data: OffboardingChecklist }>(`/hrm/offboarding/${id}/complete`)
       .then(r => r.data.data),
};
